### ISSUE TITLE
Register custom validation on struct type

### ISSUE THREAD

**Issue by iliyanmotovski** [go-playground/validator#470]
### Package version eg. v8, v9: 

v9

### Issue, Question or Enhancement:

Question

### Code sample, to showcase or reproduce:

```go
type NullBool struct {
	Value  bool
	Valid  bool
}

type User struct {
	IsSubaccount  NullBool  `validate:"isValidNullBool"`
}

var validate *validator.Validate

func init() {
	validate = validator.New()
	validate.RegisterValidation("isValidNullBool", isValidNullBool)
}

func isValidNullBool(fl validator.FieldLevel) bool {
	nb, ok := fl.Field().Interface().(NullBool)
	if !ok {
		return false
	}

	return nb.Valid
}
```
Is it possible to do custom validation on a field which is of type ```struct```? Right now the custom validator is not being called.

**Comment by deankarn** [comment]
Hey @iliyanmotovski currently the field level validation cannot be used with a struct, however, there is a struct level validation you can register in its stead, see [here](https://github.com/go-playground/validator/blob/v9/_examples/struct-level/main.go#L37) for the example.

I do hope to change this sometime in the future to have a unified validator for both, but that's a way off.

let me know if that helps :)

**Comment by iliyanmotovski** [comment]
Yes, however to my understanding, struct level validation is called per ```type```, this means that that the validation will run **_every_** time it encounters the concrete ```type``` it's registered for. However this is not the behavior I'm looking for, I want to be able to tell when to run this validation :)

**Comment by eugeneradionov** [comment]
@iliyanmotovski @deankarn I have the same situation... If I want to use **one** struct in two **different** structs and want to validate it with different rules I have no way to do this (except duplicating it with a different name and add validation rules to it).

**Example of what I need**
```go
type Product {
    Attributes `validate:"my_custom_validation"`
    // ... other fields
}

type ProductDraft {
    Attributes // no validation here
    // ... other fields
}
```

**Comment by sergolius** [comment]
@iliyanmotovski , @eugeneradionov
Here is limited possible solution, it doesn't work for deep nested structs(more than 1):

```go
import (
	"reflect"

	"gopkg.in/go-playground/validator.v9"
)

type NullBool struct {
	Value bool
	Valid bool
}

type User struct {
	IsSubaccount NullBool `validate:"isValidNullBool"`
}

var validate *validator.Validate

func init() {
	validate = validator.New()
	validate.RegisterStructValidation(isValidNullBoolStructUser, NullBool{})
}

func isValidNullBoolStructUser(sl validator.StructLevel) {
	// check that top is expected one
	okUser := indirectType(sl.Top().Type()) == reflect.TypeOf(User{}) // uniqueness based on type of Top
	if !okUser {
		return
	}

	nb := sl.Current().Interface().(NullBool)

	// produce an error
	if !nb.Valid {
		sl.ReportError(nb.Valid, "Valid", "valid", "valid", "")
	}
}

func indirectType(v reflect.Type) reflect.Type {
	for v.Kind() != reflect.Ptr {
		return v
	}
	return v.Elem()
}

```

Define function than checks type of `Top` struct of the nested.

**Comment by devstein** [comment]
@deankarn Any updates on this? I'm looking for a way to parameterize a `StructLevel` validation function via tags?

```go
type Number struct {
     Value int
}

type B struct {
   NumberOne Number `validate:"only_allow=1"`
   NumberTwo Number `validate:"only_allow=2"`
}

// I want to register a Number validator called only_allow 
// that has access to the parameters (1,2, etc)

```

**Comment by sergolius** [comment]
@devstein Did you tried `oneof` ?

**Comment by devstein** [comment]
@sergolius I checked [the code](https://github.com/go-playground/validator/blob/master/baked_in.go#L209) and it looks like `oneof` doesn't support `Struct`

**Comment by eugeneradionov** [comment]
@devstein write custom validation, for now it's the one way

**Comment by devstein** [comment]
@eugeneradionov Thanks! I used a custom type func to convert my `struct` into a primitive that `validator` supports with the built-in methods. My solution was roughly  the following:

```go
type Number struct {
     Value int
}

type B struct {
   NumberOne Number `validate:"oneof=1"`
   NumberTwo Number `validate:"oneof=2"`
}

validate.RegisterCustomTypeFunc(NumberValue, Number{})

func NumberValue(v reflect.Value) interface{} {
    n, ok := v.Interface().(Number)
     
    if !ok {
        return nil
    }

    return n.Value
}
```

**Comment by deankarn** [comment]
Sorry for the radio silence, haven’t had a lot of free time lately with everything going on, thank you for everyone for jumping in and helping :)

**Comment by rr-justin-hanselman** [comment]
I know this issue is a version too late, but for v10 this is still an issue.

I would like to float the idea here (before maybe tackling a PR).  

What I would like to see is the addition of the Tag to the StructLevel for a StructLevel Validator.

My Use Case involves Decimals:

```go
Money {
   Amount decimal.Decimal `validate:"min=0,max=500,prec=2"`
}
```

In this scenario, I would like to avoid the conversion to floats so that I can have the precision afforded by the struct, as well as use a "precision" call to check the decimal precision (which is lost if I convert to a float)

Since I don't easily have a way of getting this info from a Struct Validation, I would either have to reparse and individually cache my parent type or just avoid this all together.

**Comment by eugeneradionov** [comment]
@rr-justin-hanselman try to define custom tags as I mentioned before https://github.com/go-playground/validator/issues/470#issuecomment-607625082

Btw: I'd suggest storing money in coins and use uint for this purpose

**Comment by vtopc** [comment]
> Btw: I'd suggest storing money in coins and use uint for this purpose

Yes, but https://github.com/shopspring/decimal is made for representing money, it uses `big.Int` under the hood. So until it is converted to/from float everything is OK.

Btw: I'd suggest use int for coins - https://play.golang.org/p/_hAuQkAlCC6

**Comment by rr-justin-hanselman** [comment]
@eugeneradionov 

The only way I could see using RegisterTagNameFunc would be to abuse the altName as Namespace attribute on a StructField?

Pardon the Pseudo-code:
```go
func structTagsAsNS(fld reflect.StructField) string {
  // Handle other types that we also want to evaluate
  if reflect.Type().AssignableTo(decimal.Decimal) {
      return fld.Tag.Get("validate")
  }
  return ""
}

func DecimalValidation(sl StructLevel) {
     
     vCmds := strings.Split(sl.ns, ",")
     for i := 0; i < len(vCmds); i++ {
        vCmd := vCmds[i]
       decValue := sl.Current().Interface().(decimal.Decimal)
        if vCmd == "prec" {
            // Let's call this the decimal precision operator
            DecimalPrecisionValidation(decValue)
        } else {
           sl.Validator().Var(decValue.Float64(), vCmd)
        }
     }
}

v := validator.New()
v.RegisterStructValidation(DecimalValidation, decimal.Decimal{})
v.RegisterTagNameFunc(structTagsAsNS)
```

This might work as a hack for sure.  I had a much more complicated hack fix involving wrapping the Context functions.

That being said, unless I'm missing something, the same thing could be achieved by passing the cTag to the StructLevel and it wouldn't muddy the functional definition of a TagNameFunc.

Let me know if I misunderstood!
